File: e2.3-insert-existing-element-without-node-copy.hs
data Tree a = Empty | Bin a (Tree a) (Tree a) deriving Show

insert e Empty = Bin e Empty Empty
insert e tree@(Bin n l r) | e < n = Bin n (insert e l) r
                     | e > n = Bin n l (insert e r)
                     | otherwise = tree

input = (Bin 8 (Bin 5 (Bin 3 Empty Empty) (Bin 6 Empty Empty)) (Bin 10 (Bin 9 Empty Empty) (Bin 11 Empty (Bin 12 Empty Empty))))
output = insert 4 input
main = print $ output

File: e2.3-insert-existing-element-without-node-copy.clj
(defmulti insert' (fn [e list] (empty? list)))
(defmethod insert' true [e list] [e [] []])
(defmethod insert' false [e [x l r :as all]] (cond (< e x) [x (insert' e l) r]
                                          (> e x) [x l (insert' e r)]
                                          :else (throw (new Exception "Key is already in tree"))))
(defn insert [e l] (try (insert' e l) (catch Exception e l)))
